### ISSUE TITLE
Promregator gets stuck after timeout

### ISSUE THREAD

**Issue by eloo** [promregator/promregator#96]
## Environment
Promregator version: 0.5.7
Running as docker in CF

## Summary / Problem Statement
Promregator is able to recover from a timeout while fetching information from the CF API.
It looks like after the first timeout no other call is working again. 
After a simple restart of promregator its working again. So it looks like the API is not every time that slow but after one failure promregator needs a restart.

## Log
```
2019-04-18 07:23:35.286  WARN 18 --- [io-8080-exec-15] o.c.p.scanner.ReactiveTargetResolver     : Error on retrieving org id for org 'myorg'
   2019-04-18T09:23:35.28+0200 [APP/PROC/WEB/0] OUT java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 5000ms (and no fallback has been configured)
```

## Expected Behavior
I would like promregator to be "self-healing" after a timeout.

**Comment by eaglerainbow** [comment]
> I would like promregator to be "self-healing" after a timeout.

Promregator should be self-healing/resilient to scraping targets which time out / do not respond properly. There was an ugly bug in 0.5.6, which I fixed with 0.5.7 (see also #94). 

Your 
```
java.util.concurrent.TimeoutException
```
exactly means that the target could be reached within the configured timeframe and control of the thread is handed back to target scraping pool. 

The log snippet, unfortunately, isn't telling me much - so I am groping in the dark. Do you have something  more (logs?) for me to work with?
Which scraping mode are you using?
What does Prometheus say how many (and which kind of) targets does it still see?
Are we sure that there is still someone asking Promregator to scrape something (driver always must be Prometheus)?

**Comment by eloo** [comment]
Okay, i will try to get more logs on the occurrence.
I'm using single target scraping in this setup and discovery is called around every 30 seconds.

And the problem is that the fetch org id call is not working.. its not about a single target.. all target within this org are not working.
And yes i'm sure that prometheus is still trying to scrape, because this log are flooding the whole logs of promregator.

**Comment by eloo** [comment]
Okay, we just had this issue again and now i have a longer log for you attached. 
This is still just a snippet with 500 lines but the other 150000 lines are looking the same.. 

And i've made a new discovery there, i have some configs with an org with UPPERCASE org name and some configs with the same org lowercased... and it looks like the UPPERCASED are working and the lowercased are going to break. 
`cf orgs` is saying that UPPERCASED is correctly, so maybe its related with that?
[prom.log](https://github.com/promregator/promregator/files/3112574/prom.log)

**Comment by eaglerainbow** [comment]
Thanks for the logs - it helps partially, but unfortunately it's not sufficient yet.
But let's see what we got:

* The log you sent over spans 7 seconds, which isn't much.
* Most of the records are just the replication of what you already had posted. It means that "something" in Promregator required to know the org metadata. This can be scraping, or this can be discovery (we can't know this from the logs). This doesn't work, as the Cloud Foundry Cloud Controller (or the communication path to it) is broken. At least for each line written `java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 5000ms (and no fallback has been configured) ` a request was sent to the CFCC, which did not answer within 5 seconds (to whatever reason).
* The high number of occurrences within the same seconds tells me that there are multiple consumers asking for that information - but all of them fail.
* Very interesting to me is the RTR line at 2019-04-24T15:38:42.33+0200 which reads

```
OUT some.url - [2019-04-24T13:38:41.426+0000] "GET /singleTargetMetrics/16367186-eb8a-4b6c-9e36-6f47f25255fe/0 HTTP/1.1" 200 0 106832 "-" "Prometheus/2.8.0" ...
```
   which means that a scraping request was sent by Prometheus (some time before) and that request could be fulfilled successfully (~106kb of metrics data has been transferred successfully). This means that the scraping thread pool still must be running properly (otherwise that would not have been possible).

* The next block at 2019-04-24T15:38:48.69+0200 is also interesting:  

```
   2019-04-24T15:38:48.69+0200 [APP/PROC/WEB/0] OUT 2019-04-24 13:38:48.697  INFO 17 --- [io-8080-exec-16] o.c.p.s.R.resolveOrg                     : onSubscribe(FluxFlatMap.FlatMapMain)
   2019-04-24T15:38:48.69+0200 [APP/PROC/WEB/0] OUT 2019-04-24 13:38:48.697  INFO 17 --- [io-8080-exec-16] o.c.p.s.R.resolveSpace                   : onSubscribe(FluxFlatMap.FlatMapMain)
   2019-04-24T15:38:48.69+0200 [APP/PROC/WEB/0] OUT 2019-04-24 13:38:48.697  INFO 17 --- [io-8080-exec-16] o.c.p.s.R.resolveApplication             : onSubscribe(FluxFlatMap.FlatMapMain)
   2019-04-24T15:38:48.69+0200 [APP/PROC/WEB/0] OUT 2019-04-24 13:38:48.697  INFO 17 --- [io-8080-exec-16] o.c.p.s.R.resolveApplication             : request(unbounded)
   2019-04-24T15:38:48.69+0200 [APP/PROC/WEB/0] OUT 2019-04-24 13:38:48.697  INFO 17 --- [io-8080-exec-16] o.c.p.s.R.resolveSpace                   : request(256)
   2019-04-24T15:38:48.69+0200 [APP/PROC/WEB/0] OUT 2019-04-24 13:38:48.697  INFO 17 --- [io-8080-exec-16] o.c.p.s.R.resolveOrg                     : request(256) 
```

   It means that some request was received which triggered another set of target resolutions: for org, space *and* applicationNames. The call is coming from the block at https://github.com/promregator/promregator/blob/v0.5.7/src/main/java/org/cloudfoundry/promregator/scanner/ReactiveTargetResolver.java#L85. If not available in the cache already, this can only come from the `CFDiscoverer` which in turn is only called by the endpoint `/discovery` or the scraping endpoints. However, given the way we are caching our stuff, they should virtually never hit an empty cache. 
* Right immediately afterwards (within the same millsecond), we have another `java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 5000ms (and no fallback has been configured) ` message. This tells me that there is another discovery/scraping going on, which already had requested fetching the metadata for org&space&applications before - but still fails after 5 seconds.

From the logs I only can see that you are still in the situation that there is no connectivity to the CFCC. Promregator can't do much in that situation - it just can report the problem and reject providing data to Prometheus. However, the behavior is not resilient (in that I agree with you).

Does this go on forever -- or does the CFCC come back? How does the log then look like?

The frequency of requests Promregator fires is quite high. Do you know whether your CF platform has a DoS detection mechanism in place? Perhaps that is what the CFCC makes to shut up (because it thinks that it is under attack)?

> And i've made a new discovery there, i have some configs with an org with UPPERCASE org name and some configs with the same org lowercased... and it looks like the UPPERCASED are working and the lowercased are going to break.

It should not. I have a unit test out there which checks that this should not matter anymore. Moreover, it wouldn't explain why you are not just firing org requests like hell, but also space requests and application names (or are they in the wrong case, too?).

**Comment by eaglerainbow** [comment]
Note to myself: the issue reported in #100 cannot be the reason for sending large amounts of org requests in such a short time frame: `resolveOrg` uses a String as key for the map (see also https://github.com/promregator/promregator/blob/v0.5.7/src/main/java/org/cloudfoundry/promregator/cfaccessor/CFAccessorCache.java#L22 )

**Comment by eaglerainbow** [comment]
Eloo, could you please send me your effective Cache Configuration?
You may easily see them by looking at the log statements

```
Cache refresh timings: ....
Cache expiry timings: ....
``` 

which are posted when Promregator starts. They might come in handy during the analysis here (I am trying to create cases where you manage to fool the cache, which *might* play a role in your problem).

**Comment by eaglerainbow** [comment]
Can you see a record somewhere in your logs which reads
```
Woof woof! It appears that the connection to the Cloud Controller is gone. Trying to restart Cloud Foundry Client
```
?

**Comment by eaglerainbow** [comment]
For the record: My suspicion is that broken resolutions are not cached properly. That would explain why we had so many failed attempts of requesting the org metadata. However, with the unit test at https://github.com/promregator/promregator/blob/9ee5ccd2d2a9f8095070b3ac11deaa769d6651b1/src/test/java/org/cloudfoundry/promregator/cache/AutoRefreshingCacheMapTest.java#L255  we would show that locking works and triggers the loadingFunction only once.

This would only leave the possibility that the `Mono<?>` is the same for all the requests within a certain expiry timeframe. Additionally, with the unit test
```java
	@Test
	public void testMonoCachedError() {
		Mono<String> mono = Mono.just("1").map(v -> {
			if (v.equals("1")) {
				throw new Error("dummy");
			}
			return v;
		}).doOnError(e -> {
			System.err.println("Error was raised");
			e.printStackTrace(System.err);
		}).onErrorResume(__ -> Mono.empty())
		.cache();
		
		mono.subscribe();
		
		String result1 = mono.block(Duration.ofMillis(100));
		System.err.println("result 1: "+result1);
		String result2 = mono.block(Duration.ofMillis(100));
		System.err.println("result 2: "+result2);
	}
```
the result is:

```
Error was raised
java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 200ms (and no fallback has been configured)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:268)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:253)
	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:370)
	at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
result 1: null
result 2: null
``` 

which means that `Mono.cache()` also caches timeouts generated during processing (note that only one TimeoutException was documented and not two).

**Comment by eaglerainbow** [comment]
This even holds true, if the following unit test is implemented (note that the `onErrorResult()` has been left out):

```java
	@Test
	public void testMonoCachedError() {
		Mono<String> mono = Mono.just("1").map(v -> {
			try {
				Thread.sleep(300);
			} catch (InterruptedException e) {
				Assert.fail("unexpected interruption");
			}
			return v;
		}).timeout(Duration.ofMillis(200)).doOnError(e -> {
			System.err.println("Error was raised");
			e.printStackTrace(System.err);
		})
		.cache();
		
		mono.subscribe();
		
		try {
			String result1 = mono.block(Duration.ofMillis(100));
			System.err.println("result 1: "+result1);
		} catch (Throwable t) {
			System.err.println("First failed");
		}
		
		try {
			String result2 = mono.block(Duration.ofMillis(100));
			System.err.println("result 2: "+result2);
		} catch (Throwable t) {
			System.err.println("Second failed");
		}

	}
```

Still the result of the output is
```
Error was raised
java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 200ms (and no fallback has been configured)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:268)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:253)
	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:370)
	at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
First failed
Second failed
``` 
(having only one single stack trace dumped).

**Comment by eloo** [comment]
@eaglerainbow 
* I dont know if our platform as a DoS protection but that could be the case
* No promregator is never recovering once this error occurs
* Our caching config looks like this
```
  request.timeout.appInSpace: 5000
  request.timeout.appSummary: 5555
  request.timeout.org: 5000
  cache:
    timeout:
      org: 86400
      space: 43200
```
from the logs
```
  2019-04-25T10:44:31.36+0200 [APP/PROC/WEB/0] OUT 2019-04-25 08:44:31.367  INFO 6 --- [           main] o.c.p.cfaccessor.CFAccessorCache         : Cache refresh timings: org cache: 86400s, space cache: 43200s, app cache: 300s, app summary cache: 300s
   2019-04-25T10:44:31.36+0200 [APP/PROC/WEB/0] OUT 2019-04-25 08:44:31.367  INFO 6 --- [           main] o.c.p.cfaccessor.CFAccessorCache         : Cache expiry timings: org cache: 120s, space cache: 120s, app cache: 120s, app summary cache: 120s
   2019-04-25T10:44:31.37+0200 [APP/PROC/WEB/0] OUT 2019-04-25 08:44:31.376  INFO 6 --- [           main] o.c.p.cfaccessor.CFAccessorCache         : Cache refresh timings: org cache: 86400s, space cache: 43200s, app cache: 300s, app summary cache: 300s
   2019-04-25T10:44:31.37+0200 [APP/PROC/WEB/0] OUT 2019-04-25 08:44:31.376  INFO 6 --- [           main] o.c.p.cfaccessor.CFAccessorCache         : Cache expiry timings: org cache: 120s, space cache: 120s, app cache: 120s, app summary cache: 120s
```
* For the one working target i guess thats the app which can be scraped
* I've not seen the Woof Woof message but i also don't have full logs
* Also interesting related to the much org calls.. i can't see them in monitoring. The rate there is quite normal (around 1 org call per minute)
* Only the org name was in wrong case for the most targets, appname and space was correct

**Comment by eaglerainbow** [comment]
That's a lot of information. Please let me digest all this for a while.

Did you really enter `  request.timeout.appInSpace: xxx` into your yaml file? Or did you provide this parameter on the CLI?

**Comment by eaglerainbow** [comment]
I am not really through yet, but I would some more things which you may do that might help me understand your issue better:
* Could you please set the log level for `org.cloudfoundry.promregator.cache.AutoRefreshingCacheMap` to `DEBUG`?  (Warning: This may create many log records over time)
  * Do you see any logs which read "Loader did not provide a value for key" or similar?
  * Around the time when the problem starts, are there any suspicious (unusual) log records which read "Refreshing key "?
* Please enable our internal metrics (`promregator.metrics.internal=true`). There are multiple metrics having prefix `promregator_autorefreshingcachemap_*` there. Look at the metrics with label `promregator_autorefreshingcachemap=org`. What's going on there if everything's still fine - and how do they change as soon as promregator starts to fail?
* Do you have a rough estimate when the problem appears first? Do you see a pattern? Could it be roughly after 24hrs of runtime?

I still will continue looking into the issue. My biggest problem right now is that I do not understand how the case (upper- / lowercase) of the org's name should come into play here.

**Comment by eaglerainbow** [comment]
> Also interesting related to the much org calls.. i can't see them in monitoring. The rate there is quite normal (around 1 org call per minute)

Well, I can start to partially understand that: You have set
``` 
cache:
    timeout:
      org: 86400
```
which mainly means  that the cache shall only be refreshed every 24hrs. That's really a lot. I currently fear that there might be a situation where the org could not be resolved properly -- and due to the way reactor is doing caching itself, it may end up that also errors are memorized! 
Just for the sake of testing: Could you lower these timeouts to something smaller? For example
``` 
cache:
    timeout:
      org: 300
      space: 300
```
Having done so, does the system recover at least after 5 minutes?

What still won't fit into my picture here is that you say that

> The rate there is quite normal (around 1 org call per minute)

With such a high timeout, you virtually should only get org requests every 24hrs (assuming that your Prometheus is scraping consistently).

**Comment by eaglerainbow** [comment]
Just for the record: I upper-cased my local org name (second target) here and fired Promregator up. Initially, there is an upper-cased org-request sent out, which is answered with the same piece of information than the lower-cased one (lower-cased is the right one). In the end there are two records in the cache: one lower-cased, one upper-cased. 
After the expiry value of the org cache, the upper-cased one got deleted, as nobody was asking anymore for the wrongly cased key.

**Comment by eaglerainbow** [comment]
What's your value for `cf.cache.timeout.resolver`? Did you change it via config?

**Comment by eaglerainbow** [comment]
> And i've made a new discovery there, i have some configs with an org with UPPERCASE org name and some configs with the same org lowercased... and it looks like the UPPERCASED are working and the lowercased are going to break.
cf orgs is saying that UPPERCASED is correctly, so maybe its related with that?

Here is just one *possible* but not verified idea how that could have happened:

Given [my analysis before](https://github.com/promregator/promregator/issues/96#issuecomment-486816797), it could mean that the uppercased org in the cache was still valid and did not require refreshing yet - or had been refreshed successfully recently. The lowercased one, though, wasn't used for long. Once the CachingTargetResolver's timeout was met, a new request had to be sent - and was hit by the CF org request timeout. And that error could have been memorized in the cache. If used recurringly that broken cache record would never get deleted - but only refreshed after 24hrs... which would feel like that it "never fixes itself".

Disclaimer once again:  Just pure guess-work of mine (trying to make our some sense of all the data here).

**Comment by eaglerainbow** [comment]
> Only the org name was in wrong case for the most targets, appname and space was correct

Did you check whether the issue still occurs, if you fix the uppercase/lowercase issue to the state which the platform reports to be correct?

**Comment by eloo** [comment]
> Did you really enter request.timeout.appInSpace: xxx into your yaml file? Or did you provide this parameter on the CLI?

Yes i'm setting it in the YAML

> Could you please set the log level for org.cloudfoundry.promregator.cache.AutoRefreshingCacheMap to DEBUG?

Should be done when the error occurs i guess?

>Look at the metrics with label cache_map_name="org". What's going on there if everything's still fine - and how do they change as soon as promregator starts to fail?

The following image shows the expiry and the red lines are indicating an outage of promregator
![image](https://user-images.githubusercontent.com/2631895/56805629-fb440e80-6829-11e9-84fc-b3f9c2d39dbe.png)

> Do you have a rough estimate when the problem appears first? Do you see a pattern? Could it be roughly after 24hrs of runtime?

No I can't see pattern for the outages

> Could you lower these timeouts to something smaller?

I'm not sure if a lower org cache timeout makes sense.. because normally a org should not change that often. Or did i something misunderstand?

> With such a high timeout, you virtually should only get org requests every 24hrs (assuming that your Prometheus is scraping constantly).

Thats exactly our idea.. because org is not changing that often and we want to reduce the number of useless calls

> What's your value for cf.cache.timeout.resolver?

Default, we have it not configured yet

> Did you check whether the issue still occurs, if you fix the uppercase/lowercase issue to the state which the platform reports to be correct?

Since the change no problem at all. But this could also be luck only at the moment.

**Comment by eaglerainbow** [comment]
> I'm not sure if a lower org cache timeout makes sense.. because normally a org should not change that often. Or did i something misunderstand?

Well, I have the _suspicion_ that (to whatever reason) you come into the situation that the timeout error gets cached in your org cache. If your timeout is defined to be 24hrs, then it would also mean this error stays in there for 24hrs before the next attempt is made to fix it (as said, this is just a non-verified assumption). 

By decreasing that value to something smaller, I hope to prove that it's not the recovery mechanism itself that is having a problem, but because nobody is asking to recover, it's not done.
And yes, I agree: What I am proposing there right now is only knocking on a door.

> Thats exactly our idea.. because org is not changing that often and we want to reduce the number of useless calls

Well, in case of an error they *might* not be so useless 😁 

(on the more later more...)

**Comment by eaglerainbow** [comment]
> Yes i'm setting it in the YAML

You see me a little bit of surprised: By default, YAML doesn't like that (and takes this as a single argument). However, Spring must have implemented some magic to make also this "slang" work. Okay, go on with it 😀 

> > Could you please set the log level for org.cloudfoundry.promregator.cache.AutoRefreshingCacheMap to DEBUG?

> Should be done when the error occurs i guess?

Yes, it would be great to have that DEBUG stuff when the error occurs. 

> The following image shows the expiry and the red lines are indicating an outage of promregator

It's hard for me to understand what's going on there. (Also for me to allow thinking aloud) here's is what I get:

* Up to the first outage apparently Promregator was restarted very often; the expiry metric is never decreased by Promregator (counter) - only restarts may reset it. 
* Around the 18th there were no expiries happening - was there no scraping running? Or were the expiry value set to a very large value?
* Over 18th till 22nd there was a nice linearly increasing curve up to the point where the second outage happened.
* After the outage happened, there was no expiry happening anymore. Most likely, the record was deleted and the cache was empty. It would be interesting to see the time-parallel value of the size of the cache then...
* After some time (roughly a day or so?), expiry started again to happen - out of a sudden? What was in the cache? Also here the current size of the cache would be interesting to know.
* After some ours of expiry happening, Promregator apparently was restarted (the steep curve going to zero). 
* Afterwards, we have the same slope  of expiry until the third outage happened.

As stated above there are multiple effects visible, which are hard for me to understand. I'd also have the following questions on top of it:

* You don't have any `promregator_autorefreshingcachemap_refresh_failure`?
* Is it possible to get `promregator_autorefreshingcachemap_refresh_success` shown in parallel to the graph (perhaps in a second chart)?

> No I can't see pattern for the outages

That's bad, that makes it even harder to spot the root cause.

> Default, we have it not configured yet

Okay, that is an important piece of information.

> Since the change no problem at all. But this could also be luck only at the moment.

If it worked now, the I don't know whether I should take it as a good sign or as a bad one 😢 

> I've not seen the Woof Woof message but i also don't have full logs

Could you please set

```
cf.watchdog: true
```

and watch out for the "Woof woof" message then?
It might happen that this only makes sense, if you have set `cache.timeout.org` to some reasonable low value (300 or so).

All in all this is still quite strange to me... I still have more even more questions than I can make a pattern out of it.
Sorry, for not being able to support you better so far 😞

**Comment by eloo** [comment]
Sadly my annotations are not working anymore :D but here you can see the expiry and size
![image](https://user-images.githubusercontent.com/2631895/56879300-6e32cc80-6a58-11e9-9765-6d503c26f9bd.png)

and i can't see any failure metrics.

FYI: Still running without issues

**Comment by eaglerainbow** [comment]
Thanks for the update. The metrics partially provide some answers, but unfortunately also creates some more:

* From the size information we can see two things: There are two orgs involved; the first one is the uppercase org that is being cached; the second one is the lowercased org name, which only is cached temporarily - as long as the resolver hasn't run again. 
* However, the pattern is totally strange. As you said, you did not touch the spring property `cf.cache.timeout.resolver`. That means the default of 300 seconds will apply there. This means the resolver runs every 300 seconds anew. Latest with that point, the lowercased org name would have been requested from the cache again. 
* Requesting it from the cache means that "last used" timestamp is set to the current date/time. That value is used to determine whether a cache record is deleted. If it is the case, an "expiry" event is logged (your first image that you have attached).
* The expiry timeout is set to 120s for your org. This means that the cache record expires before next resolver run is performed. Thus, the lowercased letters would have to be "re-fetched" every 300s. This record then would stay again for 120s before it is removed again.
* During these 120s there would be a second record in the cache (the uppercased org cache record would be used every scraping request, which happens ~15-30 seconds - so no expiry takes place there). I therefore would expect to see a much more high-frequent "cache load and expire" pattern on the size metric than your image on the previous (on the left hand side) shows.
* The Refresh-Success metric is incremented each time, when a value was not displaced, but reached your configured timeout of 24hrs. We see this metric on the right hand side of your image you posted in the previous post.
* There we see that on 19th  the first refresh took place. Most likely this is for the uppercased org name only. 
* Throughout to the 23rd we see increments which could happen daily (and thus would fit to your configuration). However, on the 23rd we see two increments which are (comparably) only shortly apart (value 4 to 5 and 5 to 6). The time duration between these two bumps are too short to come from the same cache record. Moreover, we also see that in the "size" diagram of the cache, on 23rd there is a long timeframe where the size did not drop under the value of 2. The start of this block also correlates well with your red marker  (in the first image) of the second downtime. From there also the expiry handling does not trigger any displacement anymore -- as both cache records are frequently requested. 
* Shortly before the restart (see first image), expiry of at least one cache record started again. From what I can read the charts, I would say that this could correlate to the refreshing of the org cache (timeout after 24hrs). This would mean that the "missing recovery" could have been caused by a broken state stored in the cache. After the refresh then happened (and data was synchronized with the CFCC again), the broken state was overwritten with a proper one.

All this would mean that the fact that you did not have any new issue so far, is just by chance: You don't have two records in the org cache anymore (as all of them point the same cache record) -- but if that one would be jeopardized (by whatever), still you would experience the same issue. If you then have the org cache timeout still at 24 hrs, Promregator would not recover within that timeframe (but only afterwards). 

In the end this boils it down to two things:
* As already mentioned above, setting
```
cache:
    timeout:
      org: 300
```

   at least limits the impact in case you encounter your issue. 
* We have to find out why/how the broken record may get into the cache and does not get displaced.

**Comment by eaglerainbow** [comment]
I think I just managed to a small step forward: I could reproduce your error message by setting
```
cf.request.timeout.org: 5
```
which more or less makes all org requests (of the cache) run into a timeout. Then, I could prove that 

* all subsequent requests to the org request were answered by the cache, because I did not hit my breakpoint at https://github.com/promregator/promregator/blob/9ee5ccd2d2a9f8095070b3ac11deaa769d6651b1/src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFAccessorImpl.java#L231 anymore.
* the error messages in the log looked exactly as you had reported them initially. 

So, it seems to me that we really have the situation that you get an org timeout "sometimes", which then is buffered by the cache (and not displaced).

**Comment by eloo** [comment]
@eaglerainbow 
okay, that sounds like you can reproduce our issues? 
do you already have an idea to fix it? can i supporrt you with something?

**Comment by eaglerainbow** [comment]
Yes, I _think_ I am able to reproduce the symptom (and why it is bad for you) - and I have an idea where it is coming from. However, to be honest, I currently have no clue how to fix it... I definitively need some time to think about it.

**Comment by eloo** [comment]
Okay, no problem. Take your time ;)

**Comment by eaglerainbow** [comment]
Just for the record: The best idea that so far came to my mind was to extend the cleanup thread of the cache class. If the value of the map is an instance of `Mono<?>`, we could dummy-`subscribe()` to it. As all our `Mono`s in Promregator are `Mono.cache()`'d, this virtually does not come at a cost. If the subscription ends up in an error raised, we simply mark the map's entry for deletion. That deletion, however,
* should be logged to the logging facilities, and
* should be covered by an internal metric to support operational monitoring (if there are too many such attempts, you must be able to know this).

**Comment by eloo** [comment]
@eaglerainbow 
okay.. sounds a like justifiable solution to me..
also i like the idea to have this a metric too

**Comment by eaglerainbow** [comment]
I have a first version of it, but still I have the impression that it's ass-backwards.

**Comment by eaglerainbow** [comment]
I think I found a better way of doing it with the approach at branch  `issue-96-alternative`.
I have to sleep at least one night over it and then review it again.

**Comment by eaglerainbow** [comment]
I acquire a liking for the alternative solution: Although it's not a "one solution fixes it for all and forever" approach (which I usually tend to prefer), it has big advantages, because it fixes the problem there, where it originates from. Additionally, the fix is prompt - but unfortunately may be unlinked and is not free of side-effects (if you keep old references of values from the cache in memory (outside of the cache), then you may displace cache entries without need). So, the fix is not fool-proof.
The first approach might be a little more fool-proof, but totally is not prompt: it may take up to 10 seconds until it is applied. We don't know what additional harm might be caused by the broken data during that timeframe. Moreover, if someone else already fixed the cache entries for us, we still could also displace it without need (but we would only do it once - and not do this reocurringly, which _might_ happen by a fool using the alternative approach).

Both approaches are available on `master` only right now. I'd like to downport the alternative approach to `0.5.x` first. Afterwards, I intend to create another `-rc*` docker version for you, @eloo, which I would cherish if you tested it in your environment. Let's hope that this then fixes your case as well.

For the record: The promregator-internal metric which records the forced displacement of a broken cache entry is called `promregator_autorefreshingcachemap_erroneous_entry_displaced` (which is partly a misnomer, as the metric is not driven by the AutoRefreshingCacheMap itself, but by the CFAccessorCache).

**Comment by eaglerainbow** [comment]
The only thing I still don't get into my brain is the fact that after you (@eloo) fixed your upper-case/lower-case issue in the target config of the org, you don't get any broken cache entries anymore. I can't reproduce this kind of error here - so there must be something else be different on your CF platform compared to my one that I am using.

**Comment by eaglerainbow** [comment]
@eloo 
For the latter discussion (on why the issue disappeared for you), it might be interesting if you could test the following two scenario:

1. Break all your org name targets by using a wrong case. Trigger a discovery (that should already be enough for our purpose) request and look at the `promregator_cffetch_latency_bucket` metric Promregator reports at `/promregatorMetrics` for the org cache. What is the average latency CF produces now?
2. Fix all your org name targets again (using the correct case). Make sure that you have restarted Promregator and trigger the discovery again. Look at the average of `promregator_cffetch_latency_bucket` for the org cache again. 

Do these two values differ significantly?

Note to myself: We do have the issue that the timeout is triggered by the `ReactiveCFAccessorImpl` and not by the TargetResolver. That means, the fact that we have a second CFCC round trip for the real org request (with the corrected case) cannot be the culprit (it's just "not nice" for the caller of Promregator). That is why we there is no need for a "CFTargetResolver latency metric" (which we don't have so far) - its performance simply is irrelevant for this case here.

**Comment by eaglerainbow** [comment]
@eloo : Double maintenance and release of `0.5.8-rc1` has been done. I also added another fix for issue #100 into that rc1, which also could improve cache handling in your case.

As said above, I'd be interested in feedback on it, before I make it an official patch release.

**Comment by eloo** [comment]
@eaglerainbow okay that sounds really interesting. 
i've created a new setup now for testing. but can you maybe upload a jar of the new rc?

**Comment by eaglerainbow** [comment]
This is a little tricky. In the end, I created a pre-release for you to get the jar file uploaded to github. Please have a look at this [pre-release](https://github.com/promregator/promregator/releases/tag/v0.5.8-rc1).

**Comment by eloo** [comment]
okay thanks.. i will try the pre-release together with the broken config (wrong case)

**Comment by eloo** [comment]
so i got some news..
also with properly configured orgs (case is correct) we see now an outage in the 0.5.7 version. the same as before.
but in 0.5.8-rc1 i see a different error as you can see in the attachment
[logs-0.5.8.txt](https://github.com/promregator/promregator/files/3155779/logs-0.5.8.txt)

**Comment by eaglerainbow** [comment]
> i've tried to get the full error in the second attachment

Wow! The full error thing is very interesting:

``` 
2019-05-08T08:15:25.81+0200 [APP/PROC/WEB/0] OUT 2019-05-08 06:15:25.581  WARN 23 --- [nio-8080-exec-2] o.c.p.scanner.ReactiveTargetResolver     : Error on retrieving org id for org 'MyOrg'
   2019-05-08T08:15:25.81+0200 [APP/PROC/WEB/0] OUT io.netty.handler.codec.CorruptedFrameException: invalid JSON received at byte position 0: 41504920646f776e20666f72206d61696e74656e616e6365
   2019-05-08T08:15:25.81+0200 [APP/PROC/WEB/0] OUT 	at io.netty.handler.codec.json.JsonObjectDecoder.decode(JsonObjectDecoder.java:170) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
   2019-05-08T08:15:25.81+0200 [APP/PROC/WEB/0] OUT 	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
   2019-05-08T08:15:25.81+0200 [APP/PROC/WEB/0] OUT 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final] 
```

This means that the connection of the CF Cloud Controller returned garbage! It should have returned a JSON object, but in fact returned `41504920646f776e20666f72206d61696e74656e616e6365`. That's a valid hex string - and if you decode it, you'll get

``` 
API down for maintenance
```

Could it be that your CF is taking a little pause?

**Comment by eaglerainbow** [comment]
I also found https://github.com/cloudfoundry/capi-release/releases/tag/1.77.0, second line in the "highlights".

**Comment by eloo** [comment]
Ah okay.. thats interesting, and yes there was a scheduled maintenance window. But it was only some minutes and promregator has not recovered from it sadly :/

**Comment by eaglerainbow** [comment]
> But i guess now that we see the cause it should be easier to fix that or not?

I really don't know. This surprise leaves me with more questions open than answered. Here is why:

In general, now we know what we have to look for - that makes things easier for sure. However, we don't know whether this is the only deviating&influencing parameter in this game.

On the other hand, all this happens on the level of the [CF client library](https://github.com/cloudfoundry/cf-java-client), which Promregator just reuses. I would expect that such maintenance outages would be covered by this library - and the application would something like a "EndpointNotReachableIOException" or so. Also given the ominous sentence of 

> CAPI now returns a 503 error with the message "API down for maintenance" during a BBR backup

(Source: https://github.com/cloudfoundry/capi-release/releases/tag/1.77.0)

which I also take as that CAPI some time before did *not* return a 503 HTTP status code (but *may be* just reporting 200 or so), I fear that we could end up quite in a "needle in a haystack" endeavor. It's also very hard to reproduce, since I don't have a CAPI installation (running a BBR backup which is taking long enough) of a CF platform installed in the box under my desk - and I doubt I ever will 💻 . These are those integration cases, which developers usually really hate... 😭 

> But it was only some minutes and promregator has not recovered from it sadly :/

That is, in fact, really the good question. If the CFCC does only respond with "API down for maintenance" to our business requests, Promregator can't do much about it. So during that timeframe there is little I think we could change to the current behavior, which is "reporting that the CFCC is sending crap". I might be able to improve the error message a little (e.g. removing the clutter), but for the duration of the downtime itself, it boils down that Promregator is lost, if the cache is not filled. 
Yet, the question is why it did not recover after the CFCC came back. Do you see a chance to get to the logs after the "API down for maintenance" phase was over?

> i also can't find the new metric promregator_autorefreshingcachemap_erroneous_entry_displaced :/

That's a thing to note, as it implies that the cache displacement is not triggered. However, given the exception type `io.netty.handler.codec.CorruptedFrameException` being raised for "API down for maintenance", it's also not too surprising: As I am quite conservative on when to trigger displacement, I only did so far in case of we encountered a `TimeoutException` -- because that was the exception which was raised in all your logs before! This then also means that in fact we have two different problems now that we are hunting down in this very same issue here. 
Just because the symptoms are similar, it's not said that the cure of the disease is the same 🤔 Moreover, just catching the exception `io.netty.handler.codec.CorruptedFrameException` on application level also is not the best thing to do: the exception is thrown from one level below of the CF client library. Just catching might be a workaround, but it is a violation of software layering for sure. 

In short, I see the following things to bring this topic here forward:
* Report an issue at [CF client library](https://github.com/cloudfoundry/cf-java-client) both asking for guidance who should handle this case and whether they want to react on that case (which I would prefer in favor of a clean solution).
* We could catch `io.netty.handler.codec.CorruptedFrameException` as a workaround for the time being, and trigger cache displacement until then. But that is a) not a durable solution and b) we don't know whether this really fixes your issue (as we had checked for the Timeout case before, but now you came around the corner with a totally different case). 
* Still the original issue with the Timeout did not happen so far. We have no clue whether the fix provided with 0.5.8-rc1 really solved your original issue (we just learned that there is yet-another-case which made Promregator not recover). So, we also don't have a clue whether the original countermeasure ever worked so far (and made Promregator recover accordingly). 

This issue becomes more and more complicated 😢 .

BTW: I will be off for vacation for ~2 weeks starting this weekend (planned since long). I don't know yet how much time I will be spending on this issue here due to that... It begins getting a problem that I am here the "only wolf"...

**Comment by eaglerainbow** [comment]
The ticket I have created at cf-java-client was tagged with "bug". I take this that consumers of the library should not catch `io.netty.handler.codec.CorruptedFrameException` themselves.

@eloo How critical is that issue for you? Do we need to go ahead with a temporary workaround for it or shall we wait for a proper solution also supported by cf-java-client?

Could you prove that Promregator could recover from the `TimeOut` case, e.g. you got at least one `promregator_autorefreshingcachemap_erroneous_entry_displaced` reported?

**Comment by eloo** [comment]
Okay, thats sounds good. At least you have found a valid bug in the used library ;) 

The issue is atm not critical for us. Runing 0.5.7 with correct cased orgs seems to work so for. We see atm proper recovering mechanism and hadn't an outage since a week.
So we don't need a workaround yet.

But the 0.5.8 rc is mostly not recovering with wrong cased orgs. 
At least i can see some promregator_autorefreshingcachemap_erroneous_entry_displaced values. But it looks like they are static. going from 0 to 400 once and from 0 to 3 the other time. Also they are related to the space summary. See attachment
![image](https://user-images.githubusercontent.com/2631895/57832863-ecad9f00-77b8-11e9-8ca2-fc77cd90a43a.png)

**Comment by eaglerainbow** [comment]
> Okay, thats sounds good. At least you have found a valid bug in the used library ;)

Well, that has also a downside: It will take quite a while until we can finally fix that case here :disappointed: 

> So we don't need a workaround yet.

That's fine. In case that this changes, please feel free to speak up and then we check what we can do for it.

> But it looks like they are static. going from 0 to 400 once and from 0 to 3 the other time. Also they are related to the space summary. 

That is not of bad at all in my eyes:
* going from 0 to "something high" is okay, because it means that a situation was detected which was bad and immediately recovered (at various cases - which sounds logical to me to happen)
* After restart of Promregator, the situation did not happen again anymore.
* Displacement worked as with the next request, the erroneous case was fetched from the CFCC again.
* It is not said by far that this problem only happens with org-fetching first. Depending in which state of cache Promregator is, it can happen at any cache. 


>  See attachment

Are you still able to get the logs there and  still check "how" the outages happened? Do we again have the `41504920646f776e20666f72206d61696e74656e616e6365` case or was this the timeout case again?
Note that the bug of cf-java-client will only help us to fix the "API down for maintenance" case - but won't do the trick with the timeout case. If it is the latter, it might be helpful to continue chasing that one...

**Comment by eloo** [comment]
Sadly i can't see the logs from so far in the past. But as far i can see the api is in maintenance mode mostly around 3 oclock. So i would say its the maintenance window what we can see around 3 oclock.

**Comment by eaglerainbow** [comment]
I found time to create a [0.5.8-rc2](https://github.com/promregator/promregator/releases/tag/v0.5.8-rc2) which should also do cache entry displacement in case that a `io.netty.handler.codec.CorruptedFrameException` is raised by the underlying library. This is very close to what a "workaround"-fix would look like. I have my doubts whether this is an appropriate way (as already said I'd rather prefer to have a solution out of https://github.com/cloudfoundry/cf-java-client/issues/964), but it regular maintenance interval gives us a good chance to see, if the displacement really permits Promregator to recover properly.

@eloo: As I know that you have the release candidates in an own environment: Would you mind installing the v0.5.8-rc2 into your test environment to see, if it recovers over night automatically? If not, it would be interesting to know the error messages in the log Promregator is still throwing.

**Comment by eloo** [comment]
@eaglerainbow okay. i have updated to the new rc.. lets see how its working

**Comment by eloo** [comment]
@eaglerainbow the new rc looks way better 
seems like recovering is now working
![image](https://user-images.githubusercontent.com/2631895/58000537-5be90300-7ad9-11e9-8d7e-b885875b312e.png)

**Comment by eaglerainbow** [comment]
Oh yes, these graphs are "beautiful", which means that we are on the right track. 

Could you do me one favor? Check the steep jump on "5/19 ~3:00" where autorefreshmapcache_erroneous jumped from 500 and something to over 1500. Zoom in quite much. What's the duration of this phase there? Or are the edges of  this phase blurred?
Background is that if it was blurred significantly, this could be problematic for both the consumers and the CFCC...

**Comment by eloo** [comment]
here is the graph
![image](https://user-images.githubusercontent.com/2631895/58019640-48a05c80-7b06-11e9-8620-72101ad72404.png)
i guess its the maintenance window again

**Comment by eaglerainbow** [comment]
> i guess its the maintenance window again

Yes, it looks like.

> here is the graph

The graph is quite interesting into three dimensions:

* Exactly right in the middle of the CFCC downtime, you see that the space summary cache times out and wants to have a new set of data. Before that, there are no new requests, because both the org and the appsInSpace requests fail. 
* At 3:06 multiple caches were expired, but still requests came in. That gave the displacement curve a huge boost - something like 700 requests or so. It would be interesting to see how many requests really were fired towards the CFCC (I hope only a small fraction of them).
* On 3:09 the downtime was over. As soon as this was the case, no more displacements took place. That is exactly what I had hoped for.

This means: If we get a proper solution from the cf-java-client library folks, which we can adapter here in Promregator, we have won the game :grin: 

Then, let's wait for them.

**Comment by eaglerainbow** [comment]
Technical note for the record:
Both corrections are currently already on the rel-0.5 branch. This means, release of a 0.5.8 technically is blocked right now.
Note that double maintenance hasn't taken place, yet properly (details has to be checked again), so we are *not* blocked on `master` for 0.6.0.

**Comment by eaglerainbow** [comment]
Update on double maintenance for this issue:

* Double maintenance of the first issue (Timeout shall not be cached) is already double maintained to `master`.
* The second issue (with the workaround for the CFCC downtime phase) has been removed from `rel-0.5` with 572b562e2edb0726348d31aa7382fcd1855308bc and outsourced into https://github.com/promregator/promregator/tree/issue-96-cfccdowntime-workaround .

This means:
* `rel-0.5` contains the first fix for the Timeout case.
* `rel-0.5` does no longer contain the second fix (for the CFCC downtime situation).
* `master' contains the first fix for the Timeout case (double maintenance).
* `master` never contained the second fix.
* `0.5.8-rc2` is (and will be the only) version which contains the workaround for the CFCC downtime situation.

:arrow_right: `rel-0.5` is in a clean state again.

**Comment by eaglerainbow** [comment]
Latest state of this issue: We are waiting for the solution of https://github.com/cloudfoundry/cf-java-client/issues/964, which will require some sort of adoption from us.

**Comment by eloo** [comment]
@eaglerainbow i've just discovered that my promregator instance gets stuck again with the following error:
2019-05-29 12:13:12.288  WARN 21 --- [     parallel-7] o.c.p.cfaccessor.CFAccessorCache         : Timed-out entry using key <myorg> detected, which would get stuck in our org cache; displacing it now to prevent further harm

the graphs are looking like this:
![image](https://user-images.githubusercontent.com/2631895/58556478-6f544680-821c-11e9-815c-a6fa07674e65.png)

**Comment by eaglerainbow** [comment]
That's not nice, because this means that we still have an issue here.
Also the graph to the very right looks ugly: it means that Promregator is displacing broken records over and over again, but the underlying data retrieval still reports garbage. 
Could it be that the IP addresses of the Cloud Controller has rotated in the meantime? At least this was the root cause I had determined in my case. There is the (otherwise undocumented) config option `promregator.workaround.dnscache.timeout` for that in place. I have set it to 60.
Could you please give it a try?

**Comment by eloo** [comment]
i'm not sure if the ip has changed.. but at least my other setup is still working :/
but i try the dns cache timeout

**Comment by eaglerainbow** [comment]
> i'm not sure if the ip has changed.. but at least my other setup is still working :/

That does not mean much. I once had observed the following behavior while two instance of Promregator were running:
* Instance 1 of Promregator was started and DNS-queried the API endpoint and received IP1 (DNS round-robin using AWS ELB was configured there).
* Instance 2 of Promregator was started some minutes later and DNS-queried the API endpoint, receiving IP2.
* Hours later, the server at IP2 (the AWS ELB) was put into maintenance mode. It thus was removed from the DNS list. the server at IP1 was kept. 
* Connection requests instance 2 was sending to IP2 simply were rejected from now on. That is why instance 2 started to fail shortly thereafter.
* Instance 1 was still fine, as connections could still be made to IP1. 

The issue with it is that instance 2 did not requery the DNS server for a new IP address. That is a known limitation of Java's default configuration. 
The parameter `promregator.workaround.dnscache.timeout` forces Java to set a timeout for all DNS queries (regardless of the TTL of the records read). I had set the value to 60, which is measured in seconds. This value might mean that Promregator (in worst case) would keep hanging for ~60 seconds before recovering from such a DNS case.

> but i try the dns cache timeout

Yap. That would be great.
Note that for proper recovering, also setting `cf.watchdog: true` might be necessary (otherwise re-querying of the DNS might not happen).

**Comment by eloo** [comment]
@eaglerainbow okay now i'm seeing the following exception
```
o.c.p.scanner.ReactiveTargetResolver     : Error on retrieving org id for org `myorg`
org.cloudfoundry.UnknownCloudFoundryException: Unknown Cloud Foundry Exception
at org.cloudfoundry.reactor.util.ErrorPayloadMapper.lambda$null$9(ErrorPayloadMapper.java:100) ~[cloudfoundry-client-reactor-3.14.0.RELEASE.jar!/:na]
at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:118) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:67) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:106) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:262) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.FluxUsing$UsingFuseableSubscriber.onNext(FluxUsing.java:351) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.FluxFilterFuseable$FilterFuseableSubscriber.onNext(FluxFilterFuseable.java:104) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onNext(FluxPeekFuseable.java:469) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1083) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.MonoReduceSeed$ReduceSeedSubscriber.onComplete(MonoReduceSeed.java:152) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:130) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
at reactor.ipc.netty.channel.FluxReceive.terminateReceiver(FluxReceive.java:380) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
at reactor.ipc.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:204) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
at reactor.ipc.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:345) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
at reactor.ipc.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:338) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
at reactor.ipc.netty.channel.ChannelOperations.onHandlerTerminate(ChannelOperations.java:404) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
at reactor.ipc.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:631) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
at reactor.ipc.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:138) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1432) ~[netty-handler-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1199) ~[netty-handler-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1243) ~[netty-handler-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278) ~[netty-codec-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:648) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:583) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:500) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897) ~[netty-common-4.1.31.Final.jar!/:4.1.31.Final]
at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException:
Assembly trace from producer [reactor.core.publisher.MonoPeek] :
	reactor.core.publisher.Mono.checkpoint(Mono.java:1375)
	reactor.core.publisher.Mono.checkpoint(Mono.java:1325)
	org.cloudfoundry.reactor.client.v2.organizations.ReactorOrganizations.list(ReactorOrganizations.java:199)
	org.cloudfoundry.promregator.cfaccessor.ReactiveCFAccessorImpl.lambda$retrieveOrgId$1(ReactiveCFAccessorImpl.java:198)
	reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:118)
	reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:108)
	reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1083)
	reactor.core.publisher.MonoZip$ZipCoordinator.signal(MonoZip.java:247)
	reactor.core.publisher.MonoZip$ZipInner.onNext(MonoZip.java:329)
	reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:1640)
	reactor.core.publisher.MonoZip$ZipInner.onSubscribe(MonoZip.java:318)
	reactor.core.publisher.MonoJust.subscribe(MonoJust.java:54)
	reactor.core.publisher.Mono.subscribe(Mono.java:3088)
	reactor.core.publisher.MonoZip.subscribe(MonoZip.java:128)
	reactor.core.publisher.MonoMap.subscribe(MonoMap.java:55)
	reactor.core.publisher.MonoFlatMap.subscribe(MonoFlatMap.java:60)
	reactor.core.publisher.MonoTimeout.subscribe(MonoTimeout.java:81)
	reactor.core.publisher.Mono.subscribe(Mono.java:3088)
	reactor.core.publisher.FluxRetry$RetrySubscriber.resubscribe(FluxRetry.java:109)
	reactor.core.publisher.FluxRetry$RetrySubscriber.onError(FluxRetry.java:93)
	reactor.core.publisher.SerializedSubscriber.onError(SerializedSubscriber.java:114)
	reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.onError(FluxTimeout.java:192)
	reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
	reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
	reactor.core.publisher.FluxPeek$PeekSubscriber.onError(FluxPeek.java:207)
	reactor.core.publisher.FluxPeek$PeekSubscriber.onError(FluxPeek.java:207)
	reactor.core.publisher.MonoNext$NextSubscriber.onError(MonoNext.java:87)
	reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onError(MonoFlatMapMany.java:193)
	reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
	reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
	reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
	reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:135)
	reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:67)
	reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:106)
	reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:262)
	reactor.core.publisher.FluxUsing$UsingFuseableSubscriber.onNext(FluxUsing.java:351)
	reactor.core.publisher.FluxFilterFuseable$FilterFuseableSubscriber.onNext(FluxFilterFuseable.java:104)
	reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onNext(FluxPeekFuseable.java:469)
	reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1083)
	reactor.core.publisher.MonoReduceSeed$ReduceSeedSubscriber.onComplete(MonoReduceSeed.java:152)
	reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:130)
	reactor.ipc.netty.channel.FluxReceive.terminateReceiver(FluxReceive.java:380)
	reactor.ipc.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:204)
	reactor.ipc.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:345)
	reactor.ipc.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:338)
	reactor.ipc.netty.channel.ChannelOperations.onHandlerTerminate(ChannelOperations.java:404)
	reactor.ipc.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:631)
	reactor.ipc.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:138)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
	io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1432)
	io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1199)
	io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1243)
	io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
	io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:648)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:583)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:500)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462)
	io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
Error has been observed by the following operator(s):
	|_	Mono.checkpoint(ReactorOrganizations.java:199)
```

**Comment by eaglerainbow** [comment]
Uff

```
org.cloudfoundry.UnknownCloudFoundryException: Unknown Cloud Foundry Exception
```
makes you almost speechless. Seeing these error messages would not be a bad thing in general. Does Promregator recover properly?
Was this during your CFCC downtime phase?

**Comment by eloo** [comment]
sadly it was not recovering :/ 
yep it started with the downtime but the issue then persists

**Comment by eaglerainbow** [comment]
A word of warning: Note that I have just released 0.5.8. The corruptedFrame correction 3b6839fd3c6d877c314f4c4e9f6b4234c4d701bd, which was part of 0.5.8-rc2 is *not* included in the final version. I still have it on a separate branch.

**Comment by Houlistonm** [comment]
Any decision on a resolution for this issue?
We're seeing scrapes fail during nightly backups, but it does recover after the backups complete.
It would be beneficial if promegrator could "freeze" the last known configuration and continue to use it until the backups complete.

**Comment by eaglerainbow** [comment]
> It would be beneficial if promegrator could "freeze" the last known configuration and continue to use it until the backups complete.

What do you mean exactly with that?
Do you mean freezing the state of discovery?
(fwiw: We also continued the discussion at https://github.com/cloudfoundry/cf-java-client/issues/964 , which currently seems to be the better place for the question on how to detect the maintenance state).

**Comment by Houlistonm** [comment]
@eaglerainbow Yes, freeze = freezing the state of discovery.
The ASSUMPTION is ... if the CF API is down for maintenance, nobody can make changes, not developers, the the system itself, nobody.  So keep scraping the "AS-IS" and don't worry about discovery until the backups are complete.
